ledger/accountdb.go

```diff
@@ -1026,74 +1080,395 @@ func accountsPutTotals(tx *sql.Tx, totals ledgercore.AccountTotals, catchpointSt
 	return err
 }
 
-// accountsNewRound updates the accountbase and assetcreators tables by applying the provided deltas to the accounts / creatables.
-// The function returns a dbAccountData for the modified accounts which can be stored in the base cache.
-func accountsNewRound(tx *sql.Tx, updates compactAccountDeltas, creatables map[basics.CreatableIndex]ledgercore.ModifiedCreatable, proto config.ConsensusParams, lastUpdateRound basics.Round) (updatedAccounts []dbAccountData, err error) {
+func loadHoldings(stmt *sql.Stmt, eah ledgercore.ExtendedAssetHolding) (map[basics.AssetIndex]basics.AssetHolding, ledgercore.ExtendedAssetHolding, error) {
+	if len(eah.Groups) == 0 {
+		return nil, ledgercore.ExtendedAssetHolding{}, nil
+	}
+	var err error
+	holdings := make(map[basics.AssetIndex]basics.AssetHolding, eah.Count)
+	for gi, g := range eah.Groups {
+		holdings, eah.Groups[gi], err = loadHoldingGroup(stmt, g, holdings)
+		if err != nil {
+			return nil, ledgercore.ExtendedAssetHolding{}, err
+		}
+	}
+	eah.SetLoaded()
+	return holdings, eah, nil
+}
 
-	var insertCreatableIdxStmt, deleteCreatableIdxStmt, deleteByRowIDStmt, insertStmt, updateStmt *sql.Stmt
+func loadHoldingGroup(stmt *sql.Stmt, g ledgercore.AssetsHoldingGroup, holdings map[basics.AssetIndex]basics.AssetHolding) (map[basics.AssetIndex]basics.AssetHolding, ledgercore.AssetsHoldingGroup, error) {
+	if holdings == nil {
+		holdings = make(map[basics.AssetIndex]basics.AssetHolding, g.Count)
+	}
+	groupData, err := loadAssetHoldingGroupData(stmt, g.AssetGroupKey)
+	if err != nil {
+		return nil, ledgercore.AssetsHoldingGroup{}, err
+	}
+	aidx := g.MinAssetIndex
+	for i := 0; i < len(groupData.AssetOffsets); i++ {
+		aidx += groupData.AssetOffsets[i]
+		holdings[aidx] = groupData.GetHolding(i)
+	}
+	g.Load(groupData)
+	return holdings, g, nil
+}
 
-	deleteByRowIDStmt, err = tx.Prepare("DELETE FROM accountbase WHERE rowid=?")
+func loadAssetHoldingGroupData(stmt *sql.Stmt, key int64) (group ledgercore.AssetsHoldingGroupData, err error) {
+	var buf []byte
+	err = stmt.QueryRow(key).Scan(&buf)
 	if err != nil {
+		return ledgercore.AssetsHoldingGroupData{}, err
+	}
+
+	err = protocol.Decode(buf, &group)
+	return
+}
+
+func accountsNewCreate(qabi *sql.Stmt, qaei *sql.Stmt, addr basics.Address, ad basics.AccountData, proto config.ConsensusParams, updatedAccounts []dbAccountData, updateIdx int) ([]dbAccountData, error) {
+	assetsThreshold := config.Consensus[protocol.ConsensusV18].MaxAssetsPerAccount
+	var pad ledgercore.PersistedAccountData
+	if len(ad.Assets) <= assetsThreshold {
+		pad.AccountData = ad
+	} else {
+		pad.ExtendedAssetHolding.ConvertToGroups(ad.Assets)
+		pad.AccountData = ad
+		pad.AccountData.Assets = nil
+	}
+
+	var result sql.Result
+	var err error
+	for i := 0; i < len(pad.ExtendedAssetHolding.Groups); i++ {
+		result, err = qaei.Exec(pad.ExtendedAssetHolding.Groups[i].Encode())
+		if err != nil {
+			break
```

Comment:
breaking here is abit more risky, since it depends on the flow-control of the method. I would suggest returning here with the error and without updatedAccounts.
( in case of an update error, the entire transaction is being rolled back, so doing the updates won't really help )